gcp/logging.cpp

```diff
@@ -0,0 +1,642 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/** @brief Lua GCP Stackdriver wrapper implementation @file */
+
+extern "C"
+{
+#include "lauxlib.h"
+#include "lua.h"
+
+#ifdef LUA_SANDBOX
+#include <luasandbox.h>
+#include <luasandbox/heka/sandbox.h>
+#include <luasandbox_output.h>
+#endif
+
+int luaopen_gcp_logging(lua_State *lua);
+}
+
+#include <chrono>
+#include <exception>
+#include <google/logging/v2/logging.grpc.pb.h>
+#include <google/protobuf/timestamp.pb.h>
+#include <grpc++/grpc++.h>
+#include <memory>
+
+#ifdef LUA_SANDBOX
+#include "common.h"
+#endif
+
+using google::logging::v2::LoggingServiceV2;
+using google::logging::v2::WriteLogEntriesRequest;
+using google::logging::v2::WriteLogEntriesResponse;
+using google::logging::v2::ListLogEntriesRequest;
+using google::logging::v2::ListLogEntriesResponse;
+using grpc::ClientContext;
+
+static const char *mt_writer = "mozsvc.gcp.logging.writer";
+
+struct async_write_request {
+  void                    *sequence_id;
+  WriteLogEntriesRequest  *request;
+  ClientContext           ctx;
+  WriteLogEntriesResponse response;
+  grpc::Status            status;
+  std::unique_ptr<grpc::ClientAsyncResponseReader<WriteLogEntriesResponse> > rpc;
+};
+
+struct writer {
+#ifdef LUA_SANDBOX
+  const lsb_logger *logger;
+#endif
+  WriteLogEntriesRequest                  *request;
+  std::unique_ptr<LoggingServiceV2::Stub> stub;
+  int                                     batch_size;
+  int                                     max_async_requests;
+  int                                     outstanding_requests;
+  grpc::CompletionQueue                   cq;
+};
+
+typedef struct writer_wrapper
+{
+  writer *w;
+} writer_wrapper;
+
+
+static int writer_new(lua_State *lua)
+{
+  int n = lua_gettop(lua);
+  luaL_argcheck(lua, n >= 1 && n <= 3, n, "incorrect number of arguments");
+  const char *channel = luaL_checkstring(lua, 1);
+  int max_async       = luaL_optint(lua, 2, 20);
+  int batch_size      = luaL_optint(lua, 3, 1000);
+
+  writer_wrapper *ww = static_cast<writer_wrapper *>(lua_newuserdata(lua, sizeof*ww));
+  ww->w = new struct writer;
+  if (!ww->w) return luaL_error(lua, "memory allocation failed");
+
+  ww->w->max_async_requests = max_async;
+  ww->w->batch_size = batch_size;
+  ww->w->outstanding_requests = 0;
+  ww->w->request = new WriteLogEntriesRequest;
+#ifdef LUA_SANDBOX
+  lua_getfield(lua, LUA_REGISTRYINDEX, LSB_THIS_PTR);
+  lsb_lua_sandbox *lsb = reinterpret_cast<lsb_lua_sandbox *>(lua_touserdata(lua, -1));
+  lua_pop(lua, 1); // remove this ptr
+  if (!lsb) {
+    return luaL_error(lua, "invalid " LSB_THIS_PTR);
+  }
+  ww->w->logger = lsb_get_logger(lsb);
+#endif
+  luaL_getmetatable(lua, mt_writer);
+  lua_setmetatable(lua, -2);
+
+  bool err = false;
+  try {
+    auto creds = grpc::GoogleDefaultCredentials();
```

Comment:
Is this instance default? Should this support service account JSON files and the like?

Reply:
It works for both: https://grpc.io/docs/guides/auth.html#using-google-token-based-authentication